Spravo/Screens/AuthFlow/FetchPhoneContacts/FetchPhoneContactsViewModel.swift

```diff
@@ -10,47 +10,71 @@ import UIKit
         self.coordinator = coordinator
         self.serviceHolder = serviceHolder
-        self.addressBookProvider = serviceHolder.get(by: AddressBookProvider.self)
+        self.contactsProvider = serviceHolder.get(by: ContactsProvider.self)
         self.phoneContactsProvider = serviceHolder.get(by: PhoneContactsProvider.self)
         self.firebaseAgent = serviceHolder.get(by: FirebaseAgent.self)
     }
-    
-    func fetchPhonesContacts(completion: @escaping (_ access: Bool) -> Void) {
-        guard let userFbId = addressBookProvider.userModel.userFacebookID,
-            !phoneContactsProvider.isPhoneContactsLoadedAlready(userFbId: userFbId) else {
-            completion(true)
-            return
-        }
-        phoneContactsProvider.requestAccess { [weak self] (result) in
-            guard let _ = self else { return }
-            if !result {
-                completion(false)
-                return
+
+    func fetchPhonesContacts(completion: @escaping (BoolResult) -> ()) {
+        //TODO(Serhii K.) question to Seniour ( Should implement [weak self]? in this block )
+        phoneContactsProvider.fetchExistingPhoneContacts { (result, contacts) in
+            switch result {
+            case .failure, .success(false):
+                completion(result)
+            case .success(true):
+                if let contacts = contacts {
+                    self.phoneContacts = contacts
+                    completion(result)
+                } else {
+                    let error = NSLocalizedString("ImportPhoneContacts.ErrorFetchingPhoneContactsFailed", comment: "Message about fetching phones contacts failed") + ": " + supportEmail
+                    completion(.failure(error))
+                }
             }
-            completion(true)
         }
     }
     
-    func syncingContacts(completion: @escaping (_ access: Bool) -> Void) {
-        //TODO(Serhii K.) will remove in next stage (realizing fetch phones contacts)
-        DispatchQueue.main.asyncAfter(deadline: .now() + TimeInterval(2)) { [weak self] in
-            guard let _ = self else { return }
-            completion(true)
+    func syncingContacts(completion: @escaping (_ error: String?) -> Void) {
+        guard let contacts = phoneContacts, let userFbId = contactsProvider.userModel.facebookId else {
+            let error = NSLocalizedString("ImportPhoneContacts.ErrorSyncingContactsFailed", comment: "Message about syncing contacts failed") + ": " + supportEmail
+            completion(error)
+            return
+        }
+        var contactsQuantity = contacts.count
+        for item in contacts {
+            var image: UIImage?
+            if let data = item.1 {
+                image = UIImage(data: data)
+            }
+            firebaseAgent.saveNewContact(userFbId: userFbId, contact: item.0, userProfileImage: image) { [weak self] error in
+                guard let self = self else { return }
```

Comment:
This block of code will work for every contact upload
If we have 50 contacts it will trigger completion 50 times, it will mark upload 50 times, it will reset phoneContacts to be [] 50 times

stange